File: sdk/python/ramp_sdk/wire_canon.py
"""Wire-to-canonical offer conversion (the from-wire canonicalizer).

The RAMP Connect wire is snake_case proto-JSON with ``EmitUnpopulated``: zero-valued
scalars, empty repeateds, null messages and ``*_UNSPECIFIED`` enums are all present.
That is what ``sdk/go/connectserver/codec.go`` marshals (``UseProtoNames: true``), and
what ``codec_test.go`` guards — a stray ``UseProtoNames=false`` is the regression it
catches. The offer SIGNATURE covers the CANONICAL form: the same snake_case proto
names, but omit-unpopulated, enums-as-names (sdk/go helpers
``canonicalSignJSONOptions``), then RFC 8785 JCS. Both sides share the naming, so what
this inversion actually undoes is the zero-inflation and the signature fields, NOT the
naming. Every Python client verifying a wire offer must still perform it before calling
:func:`ramp_sdk.core.canonical_offer_payload` — never call that on raw wire input.

The retired proto-JSON lowerCamel ``json_name`` form is still TOLERATED on input (hence
:func:`_snake`, the ``offer_camel`` parameter name, and the pre-flip fixture the tests
keep). It is out of contract: accepted when it arrives, never emitted.

:func:`from_wire_offer` performs that inversion SCHEMA-AWARE, driven by the
generated ``wire.models`` classes (gen/python), whose field DEFAULTS encode
proto3 presence: ``default None`` marks a presence-tracked field (proto3
optional scalar, message, repeated, map) whose wire presence means SET (kept
verbatim — including an optional string deliberately set to ``""``), while a
zero default (``""``/``0``) or a required field marks a non-optional scalar
whose zero value the canonical form omits (dropped). Repeated fields drop when
empty; unset messages/Structs arrive as ``null`` and drop; a set-but-empty
message (``{}``) is kept; map values keep their keys VERBATIM (map keys are
data, never case-converted). ``*_UNSPECIFIED`` enum values are zero values and
drop.

Byte-parity with the Go oracle is pinned by ``tests/test_wire_canon.py``: over the
drift-gated ``sdk/go/helpers/testdata/wire-canonical-vectors.json`` corpus for the
current snake_case wire, and over a live-captured fixture pair for the retired form.

KNOWN LIMITS (fail-closed, never fail-open): a wire offer carrying fields newer
than the pinned gen models is kept verbatim and will verify FALSE (rejected,
not silently accepted), and a non-optional int64 zero (emitted as ``"0"``) is
not dropped (no such field exists on the Offer tree today).
"""

from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any, get_args, get_origin

from wire import base as wire_base
from wire import models as wire_models

if TYPE_CHECKING:
    from pydantic.fields import FieldInfo

_CAMEL_BOUNDARY = re.compile(r"(?<=[a-z0-9])([A-Z])")
_UNSPECIFIED_ENUM = re.compile(r"^[A-Z][A-Z0-9_]*_UNSPECIFIED$")


def _snake(key: str) -> str:
    """Invert protojson's retired lowerCamel json_name back to the proto name.

    Idempotent on already-snake keys, which is what the current wire and the shared
    corpus emit — so the inversion costs nothing and keeps the retired form working.
    """
    return _CAMEL_BOUNDARY.sub(r"_\1", key).lower()


# Sentinel: the field would be omitted entirely by the canonical marshal.
_OMIT = object()


def _message_arg(annotation: Any) -> type[Any] | None:
    """Return the WireModel subclass inside an annotation, if any."""
    if isinstance(annotation, type) and issubclass(annotation, wire_base.WireModel):
        return annotation
    for arg in get_args(annotation):
        found = _message_arg(arg)
        if found is not None:
            return found
    return None


def _is_map(annotation: Any) -> bool:
    """True when the annotation is a proto map/Struct (a plain dict type)."""
    if get_origin(annotation) is dict:
        return True
    return any(_is_map(arg) for arg in get_args(annotation))


def _canon_scalar(value: Any, *, presence: bool) -> Any:
    """Canonicalize a scalar field value per proto3 presence semantics."""
    if presence:
        return value  # set (it is on the wire) → part of the signed bytes
    zeroish = value == "" or value is False or (not isinstance(value, bool) and value == 0)
    if zeroish or (isinstance(value, str) and _UNSPECIFIED_ENUM.match(value)):
        return _OMIT
    return value


def _canon_field(field: FieldInfo, value: Any) -> Any:
    """Canonicalize one wire field value using the gen model's field metadata."""
    if value is None:
        return _OMIT  # unset message/Struct (EmitUnpopulated renders null)
    msg_cls = _message_arg(field.annotation)
    if isinstance(value, list):
        items = [
            _canon_message(msg_cls, v) if msg_cls and isinstance(v, dict) else v for v in value
        ]
        return items or _OMIT  # canonical omits empty repeated
    if isinstance(value, dict):
        if msg_cls is not None:
            return _canon_message(msg_cls, value)  # set-but-empty {} stays
        if _is_map(field.annotation):
            return value or _OMIT  # map keys stay VERBATIM
        return value
    return _canon_scalar(value, presence=field.default is None)


def _canon_message(model_cls: type[Any], wire: dict[str, Any]) -> dict[str, Any]:
    """Reconstruct one message's canonical dict from its wire dict."""
    fields = model_cls.model_fields
    out: dict[str, Any] = {}
    for key, value in wire.items():
        name = _snake(key)
        field = fields.get(name)
        if field is None:
            # Newer-than-pin field: keep it so verification fails CLOSED (the
            # signature covered bytes this pin cannot reconstruct).
            out[name] = value
            continue
        canon = _canon_field(field, value)
        if canon is not _OMIT:
            out[name] = canon
    return out


def from_wire_offer(offer_camel: dict[str, Any]) -> dict[str, Any]:
    """Reconstruct the canonical (signed) offer dict from the wire offer dict."""
    return _canon_message(wire_models.Offer, offer_camel)

File: sdk/ts/core/wire-canon.ts
// sdk/ts/core — the wire→canonical offer inversion (the from-wire canonicalizer).
// Companion to the Verifier's canonicalOfferPayload: that face clears the two
// signature fields and JCS-es an ALREADY-canonical offer; THIS face reconstructs
// the canonical form from the RAMP Connect WIRE offer (proto3-JSON with
// EmitUnpopulated — zero-valued scalars, empty repeateds, null messages, and
// *_UNSPECIFIED enums all present). Every TS client verifying a wire offer must
// invert the wire emission BEFORE canonicalOfferPayload, never call it on raw wire.
//
// Mirror of sdk/python ramp_sdk.wire_canon.from_wire_offer. Its presence rules are
// SCHEMA-AWARE, driven by the generated gen/ts OfferSchema exactly as the Python
// side reads them from the gen/python model field defaults:
//
//   .optional()  → proto3 presence-tracked → a set-but-zero scalar is KEPT
//                  (an optional string deliberately "" survives, e.g. pricing.unit).
//   .default("")/.default(0)/bare-required → non-optional → the zero value is DROPPED.
//   z.object     → message: recurse (a set-but-empty {} stays, a null message drops).
//   z.array      → repeated: drop when empty, else canonicalize each element.
//   z.record     → map/Struct: keep VERBATIM (keys are data, never pruned or cased),
//                  drop only when empty.
//   *_UNSPECIFIED enum values are zero values and drop.
//
// Byte-parity with the Go oracle is pinned by tests/wire-canonical.parity.test.ts
// over the shared sdk/go/helpers/testdata/wire-canonical-vectors.json corpus.
//
// FAIL-CLOSED: a wire field newer than the pinned OfferSchema is kept verbatim, so a
// signature covering bytes this pin cannot reconstruct verifies FALSE (rejected),
// never silently accepted.

import { OfferSchema } from "../../../gen/ts/wire/schemas.ts";

// A structural view of the pieces of a Zod schema this inversion reads — enough to
// walk the generated OfferSchema tree without coupling to the zod value import.
interface ZodDef {
	readonly typeName: string;
	readonly innerType?: AnyZod;
	readonly type?: AnyZod;
	readonly schema?: AnyZod;
}
interface AnyZod {
	readonly _def: ZodDef;
}

// Sentinel: this field is omitted entirely by the canonical marshal.
const OMIT: unique symbol = Symbol("ramp.core.wire-canon.omit");

const CAMEL_BOUNDARY = /([a-z0-9])([A-Z])/g;
const UNSPECIFIED_ENUM = /^[A-Z][A-Z0-9_]*_UNSPECIFIED$/;

// Wrapper node types that decorate a field without changing its wire identity.
const WRAPPERS = new Set(["ZodOptional", "ZodDefault", "ZodNullable", "ZodBranded", "ZodEffects"]);

function zdef(schema: AnyZod): ZodDef {
	return schema._def;
}

// innerOf peels one wrapper: ZodOptional/ZodDefault/ZodNullable expose `innerType`,
// ZodArray/ZodBranded expose `type`, ZodEffects exposes `schema` — mutually exclusive.
function innerOf(schema: AnyZod): AnyZod {
	const d = zdef(schema);
	const inner = d.innerType ?? d.type ?? d.schema;
	if (inner === undefined) throw new Error("ramp/core: zod wrapper has no inner type");
	return inner;
}

function shapeOf(schema: AnyZod): Record<string, AnyZod> {
	return (schema as unknown as { shape: Record<string, AnyZod> }).shape;
}

// coreType strips presence/decoration wrappers to the underlying kind (ZodObject,
// ZodArray, ZodRecord, or a scalar). It STOPS at ZodArray so a repeated field is
// still detectable as repeated (the element is unwrapped separately).
function coreType(schema: AnyZod): AnyZod {
	let s = schema;
	while (WRAPPERS.has(zdef(s).typeName)) s = innerOf(s);
	return s;
}

// isPresenceTracked mirrors Python's `field.default is None`: a `.optional()` field
// is proto3 presence-tracked (keep its zero value); `.default(...)` or a bare
// required field is non-optional (drop its zero value).
function isPresenceTracked(schema: AnyZod): boolean {
	const t = zdef(schema).typeName;
	if (t === "ZodOptional") return true;
	if (t === "ZodDefault") return false;
	if (t === "ZodNullable" || t === "ZodBranded" || t === "ZodEffects") {
		return isPresenceTracked(innerOf(schema));
	}
	return false;
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

// snake inverts protojson's lowerCamel json_name back to the proto name. It is
// idempotent on already-snake keys (the shared corpus is snake_case wire).
function snake(key: string): string {
	return key.replace(CAMEL_BOUNDARY, "$1_$2").toLowerCase();
}

function canonScalar(value: unknown, presence: boolean): unknown {
	if (presence) return value; // set (on the wire) → part of the signed bytes
	const zeroish = value === "" || value === false || (typeof value === "number" && value === 0);
	if (zeroish) return OMIT;
	if (typeof value === "string" && UNSPECIFIED_ENUM.test(value)) return OMIT;
	return value;
}

function canonField(field: AnyZod, value: unknown): unknown {
	if (value === null) return OMIT; // unset message/Struct (EmitUnpopulated renders null)
	const core = coreType(field);
	const kind = zdef(core).typeName;
	if (Array.isArray(value)) {
		const elem = kind === "ZodArray" ? coreType(innerOf(core)) : undefined;
		const elemIsMessage = elem !== undefined && zdef(elem).typeName === "ZodObject";
		const items = value.map((v) =>
			elemIsMessage && elem !== undefined && isPlainObject(v) ? canonMessage(elem, v) : v,
		);
		return items.length > 0 ? items : OMIT; // canonical omits empty repeated
	}
	if (isPlainObject(value)) {
		if (kind === "ZodObject") return canonMessage(core, value); // set-but-empty {} stays
		if (kind === "ZodRecord") return Object.keys(value).length > 0 ? value : OMIT; // map keys verbatim
		return value;
	}
	return canonScalar(value, isPresenceTracked(field));
}

// keep writes a member the canonical form must carry, including one whose name is an
// inherited property of every object. A plain assignment to "__proto__" replaces the
// object's prototype and creates no member; defineProperty creates the member, which is
// what JSON.stringify — and therefore JCS, and therefore the signature — reads.
function keep(out: Record<string, unknown>, name: string, value: unknown): void {
	Object.defineProperty(out, name, {
		value,
		enumerable: true,
		writable: true,
		configurable: true,
	});
}

function canonMessage(objectSchema: AnyZod, wire: Record<string, unknown>): Record<string, unknown> {
	const shape = shapeOf(objectSchema);
	const out: Record<string, unknown> = {};
	for (const [key, value] of Object.entries(wire)) {
		const name = snake(key);
		// hasOwn on the READ, and defineProperty on the WRITE. Both are the same key: a
		// wire name like "__proto__" or "constructor" resolves to an inherited member of
		// Object.prototype, so a truthiness test hands the walk an object with no _def,
		// and a plain assignment invokes the prototype SETTER instead of creating a
		// member. An offer comes from a peer, so the key is attacker-chosen, and the
		// second half was the dangerous one: the member vanished from the canonical form
		// while the object silently inherited whatever the attacker put there. The signed
		// bytes were unaffected, so an offer carrying an appended __proto__ still VERIFIED
		// — the one member name that could be added to a signed offer for free — and the
		// caller then read attacker-chosen values off a VerifiedOffer.
		const field = Object.hasOwn(shape, name) ? shape[name] : undefined;
		if (field === undefined) {
			// Newer-than-pin field: keep it so verification fails CLOSED (the signature
			// covered bytes this pin cannot reconstruct).
			keep(out, name, value);
			continue;
		}
		const canon = canonField(field, value);
		if (canon !== OMIT) keep(out, name, canon);
	}
	return out;
}

/**
 * fromWireOffer reconstructs the canonical (signed) offer object from the WIRE
 * offer object, mirroring sdk/python from_wire_offer. Feed its result to
 * canonicalOfferPayload (or RFC 8785 JCS directly) to reproduce the exact bytes the
 * offer signature covers.
 */
export function fromWireOffer(offerWire: Record<string, unknown>): Record<string, unknown> {
	return canonMessage(OfferSchema as unknown as AnyZod, offerWire);
}
